datumaro/plugins/splitter.py

```diff
@@ -0,0 +1,480 @@
+    - If there are not enough images in some class or attributes group,
+      the split ratio can't be guaranteed.|n
+    """
+
+    def __init__(self, dataset:Dataset, splits, seed=None):
+        """
+        Parameters
+        ----------
+        dataset : Dataset
+        splits : list
+            A list of (subset(str), ratio(float))
+            Subset is expected to be one of ["train", "val", "test"].
+            The sum of ratios is expected to be 1.
+        seed : int, optional
+        """
+        super().__init__(dataset, seed)
+
+        subsets, sratio = self._validate_splits(splits)
+
+        ## support only single label for a DatasetItem
+        ## 1. group by label
+        by_labels = dict()
+        for idx, item in enumerate(self._extractor):
+            labels = []
+            for ann in item.annotations:
+                if ann.type == AnnotationType.label:
+                    labels.append(ann)
+            assert len(labels) == 1, \
+                "Expected exact one label for a DatasetItem"
+            ann = labels[0]
+            if not hasattr(ann, 'label') or ann.label is None:
+                label = None
+            else:
+                label = str(ann.label)
+            if label not in by_labels:
+                by_labels[label] = []
+            by_labels[label].append((idx, ann))
+
+        by_splits = dict()
+        for subset in self._subsets:
+            by_splits[subset] = []
+
+        ## 2. group by attributes
+        required = self._get_required(sratio)
+        for label, items in by_labels.items():
+            np.random.shuffle(items)
+            by_attributes = self._group_by_attributes(items)
+            for attributes, indice in by_attributes.items():
+                gname = 'label: {}, attributes: {}'.format(label, attributes)
+                splits = self._split_indice(indice, gname, sratio, required)
+                for subset, split in zip(subsets, splits):
+                    if len(split) > 0:
+                        by_splits[subset].extend(split)
+
+        for subset in self._subsets:
+            self._parts.append((set(by_splits[subset]), subset))
+
+class SplitforMatchingReID(TaskSpecificSplitter):
+    """
+    Splits dataset for matching, especially re-id task.|n
+    First, splits dataset into 'train+val' and 'test' sets by person id.|n
+    Note that this splitting is not by DatasetItem. |n
+    Then, tags 'test' into 'gallery'/'query' in class-wise random manner.|n
+    Then, splits 'train+val' into 'train'/'val' sets in the same way.|n
+    Therefore, the final subsets would be 'train', 'val', 'test'. |n
+    And 'gallery', 'query' are tagged using anntoation group.|n
+    You can get the 'gallery' and 'query' subsets using 'get_subset_by_group'.|n
+    Notes:|n
+    - Single label is expected for each DatasetItem.|n
+    - Each label is expected to have attribute representing the person id. |n
+    """
+    _group_map = dict()
```

Comment:
This will be a global variable at class level.

Reply:
Yes. it is used by 'get_subset_by_group' function, so I make it a global variable.